src/net.cpp

```diff
@@ -366,7 +366,16 @@ bool CConnman::CheckIncomingNonce(uint64_t nonce)
 {
     LOCK(m_nodes_mutex);
     for (const CNode* pnode : m_nodes) {
-        if (!pnode->fSuccessfullyConnected && !pnode->IsInboundConn() && pnode->GetLocalNonce() == nonce)
+        // Omit private broadcast connections from this check to prevent this privacy attack:
```

Comment:
> The peer starts connecting to (clearnet) nodes and sends them a VERSION message which contains our nonce. If the peer manages to connect to us we would disconnect.

I'm confused by this attack. First, if the peer connects to us over clearnet it would be an inbound connection for us - we don't disconnect inbounds, see `!pnode->IsInboundConn()` a few lines below. Second, even if we did disconnect, I still don't see the how the attack could work - it would be a new connection with a new ID and a new nonce (the nonce is created using the peer id [here](https://github.com/bitcoin/bitcoin/blob/d04324a7056a735c1127ba8ccdc720a16e7281a3/src/net.cpp#L509), so it should be different for each peer).

Reply:
I reread your statement and the code a few times and I think that the code is legit (or the attack the code is guarding against could happen). Let me elaborate:

1. We connect for a private broadcast to `boopeer.onion` and send them `nonce=1234` in our `VERSION` message. For this an entry is added to `CConnman::m_nodes` that has `IsInboundConn() == false`.
2. The `boo` guy delays their `VERACK` on the private connection and connects to random nodes on clearnet. Assuming we also listen on clearnet address, they are lucky enough and manage to connect to us. They send us `VERSION` with `nonce=1234`. For this connection `IsInboundConn() == true` on our side and this code is executed:

https://github.com/bitcoin/bitcoin/blob/c8e3978114716bb8fb10695b9d187652f3ab4926/src/net_processing.cpp#L3532-L3537

passing `nonce=1234` to `CheckIncomingNonce()` which finds the entry in `m_nodes` from 1. and returns `false`. Then the clearnet connection from 2. is closed by us.

Makes sense?

Reply:
Yes, now it makes sense, thanks! I'm not sure where I got confused, I probably misread the existing code.